token-erc-721/chaincode-javascript/lib/tokenERC721.js

```diff
@@ -0,0 +1,403 @@
+        const iterator = await ctx.stub.getStateByPartialCompositeKey(nftPrefix, []);
+
+        // Count the number of returned composite keys
+        let totalSupply = 0;
+        let result = await iterator.next();
+        while (!result.done) {
+            totalSupply++;
+            result = await iterator.next();
+        }
+        return totalSupply;
+    }
+
+    // ============== Extended Functions for this sample ===============
+
+    /**
+     * Set optional information for a token.
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} name The name of the token
+     * @param {String} symbol The symbol of the token
+     */
+    async SetOption(ctx, name, symbol) {
+
+        // Check minter authorization - this sample assumes Org1 is the issuer with privilege to set the name and symbol
+        const clientMSPID = ctx.clientIdentity.getMSPID();
+        if (clientMSPID !== 'Org1MSP') {
+            throw new Error('client is not authorized to set the name and symbol of the token');
+        }
+
+        await ctx.stub.putState(nameKey, Buffer.from(name));
+        await ctx.stub.putState(symbolKey, Buffer.from(symbol));
+        return true;
+    }
+
+    /**
+     * Mint a new non-fungible token
+     *
+     * @param {Context} ctx the transaction context
+     * @param {String} tokenId Unique ID of the non-fungible token to be minted
+     * @param {String} tokenURI URI containing metadata of the minted non-fungible token
+     * @returns {Object} Return the non-fungible token object
+    */
+    async MintWithTokenURI(ctx, tokenId, tokenURI) {
+
+        // Check minter authorization - this sample assumes Org1 is the issuer with privilege to mint a new token
+        const clientMSPID = ctx.clientIdentity.getMSPID();
+        if (clientMSPID !== 'Org1MSP') {
+            throw new Error('client is not authorized to mint new tokens');
+        }
+
+        // Get ID of submitting client identity
+        const minter = ctx.clientIdentity.getID();
+
+        // Check if the token to be minted does not exist
+        const exists = await this._nftExists(ctx, tokenId);
+        if (exists) {
+            throw new Error(`The token ${tokenId} is already minted.`);
+        }
+
+        // Add a non-fungible token
+        const tokenIdInt = parseInt(tokenId);
+        if (isNaN(tokenIdInt)) {
+            throw new Error(`The tokenId ${tokenId} is invalid. tokenId must be an integer`);
+        }
+        const nft = {
+            tokenId: tokenIdInt,
+            owner: minter,
+            tokenURI: tokenURI
+        };
+        const nftKey = ctx.stub.createCompositeKey(nftPrefix, [tokenId]);
+        await ctx.stub.putState(nftKey, Buffer.from(JSON.stringify(nft)));
+
+        // Save a composite key to count the balance of an owner
+        // Only the key name is stored to count the balance by getStateByPartialCompositeKey()
```

Comment:
This comment is not clear to me.
I would provide an example, e.g. composite key would be balancePrefix.owner.tokenId, which enables partial composite key query to find and count all records matching balance.owner.*